File: flutter/elvish_mail/lib/crypto/protected_link_crypto.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';

import 'account_wrap.dart';

/// Mode B protected-link ciphertext (parity with `static/mail/compose.jsx`).
class ProtectedLinkPayload {
  ProtectedLinkPayload({
    required this.kdf,
    required this.kdfSaltB64,
    required this.kdfParamsJson,
    required this.wrappedMsgKeyB64,
    required this.bodyCiphertextB64,
  });

  final String kdf;
  final String kdfSaltB64;
  final String kdfParamsJson;
  final String wrappedMsgKeyB64;
  final String bodyCiphertextB64;
}

class ProtectedLinkCryptoException implements Exception {
  ProtectedLinkCryptoException(this.message);
  final String message;
  @override
  String toString() => message;
}

Future<ProtectedLinkPayload> buildProtectedLinkPayload({
  required String from,
  required String subject,
  required String body,
  required String password,
}) async {
  final trimmedPw = password.trim();
  if (trimmedPw.length < 12) {
    throw ProtectedLinkCryptoException('Password must be at least 12 characters.');
  }
  final fullBody = 'From: ${from.isEmpty ? 'anonymous' : from}\r\n'
      'Subject: ${subject.isEmpty ? '(no subject)' : subject}\r\n\r\n$body';
  final plain = utf8.encode(fullBody);

  final msgKey = ElvishAccountWrap.randomBytes(32);
  final msgNonce = ElvishAccountWrap.randomBytes(12);
  final aes = AesGcm.with256bits();
  final symKey = SecretKey(msgKey);
  final secretBox = await aes.encrypt(plain, secretKey: symKey, nonce: msgNonce);
  final payloadCt = Uint8List.fromList([
    ...msgNonce,
    ...secretBox.cipherText,
    ...secretBox.mac.bytes,
  ]);

  final salt = ElvishAccountWrap.randomBytes(16);
  final derived = await ElvishAccountWrap.deriveKekWithMetadata(
    password: trimmedPw,
    salt: salt,
  );
  final wrapped = await ElvishAccountWrap.aesWrap(kek: derived.kek, plaintext: msgKey);

  return ProtectedLinkPayload(
    kdf: derived.kdf,
    kdfSaltB64: base64Encode(salt),
    kdfParamsJson: derived.paramsJson,
    wrappedMsgKeyB64: base64Encode(wrapped),
    bodyCiphertextB64: base64Encode(payloadCt),
  );
}

File: IOS/IOS/Mail/ProtectedLinkCrypto.swift
import CryptoKit
import Foundation

/// Client-side protected-link payload (Mode B; parity with `static/mail/compose.jsx` `sendProtectedLink`).
enum ProtectedLinkCrypto {
    struct Payload: Sendable {
        let kdf: String
        let kdfSaltB64: String
        let kdfParamsJSON: String
        let wrappedMsgKeyB64: String
        let bodyCiphertextB64: String
    }

    static func buildPayload(from: String, subject: String, body: String, password: String) throws -> Payload {
        let trimmedPw = password.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmedPw.count >= 12 else {
            throw ProtectedLinkError.passwordTooShort
        }
        let fullBody = "From: \(from.isEmpty ? "anonymous" : from)\r\nSubject: \(subject.isEmpty ? "(no subject)" : subject)\r\n\r\n\(body)"
        let plain = Data(fullBody.utf8)

        var msgKey = try ElvishAccountWrap.randomBytes(32)
        defer { msgKey.resetBytes(in: 0 ..< msgKey.count) }

        let msgNonce = try ElvishAccountWrap.randomBytes(12)
        let symKey = SymmetricKey(data: msgKey)
        let sealed = try AES.GCM.seal(plain, using: symKey, nonce: AES.GCM.Nonce(data: msgNonce))
        guard let combined = sealed.combined else {
            throw ProtectedLinkError.encryptFailed
        }
        // `combined` is nonce || ciphertext || tag (same layout as web `msgNonce` + subtle.encrypt output).
        let payloadCt = combined

        let salt = try ElvishAccountWrap.randomBytes(16)
        let derived = try ElvishAccountWrap.deriveKEKWithMetadata(
            password: trimmedPw,
            salt: salt,
            kdf: "argon2id",
            kdfParamsJSON: nil
        )
        let wrapped = try ElvishAccountWrap.aesWrap(kek: derived.kek, plaintext: msgKey)

        return Payload(
            kdf: derived.kdf,
            kdfSaltB64: salt.base64EncodedString(),
            kdfParamsJSON: derived.paramsJSON,
            wrappedMsgKeyB64: wrapped.base64EncodedString(),
            bodyCiphertextB64: payloadCt.base64EncodedString()
        )
    }
}

enum ProtectedLinkError: Error, LocalizedError {
    case passwordTooShort
    case encryptFailed

    var errorDescription: String? {
        switch self {
        case .passwordTooShort:
            return "Password must be at least 12 characters."
        case .encryptFailed:
            return "Could not encrypt protected link payload."
        }
    }
}
